pkg/ccl/backupccl/backup_compaction.go

```diff
@@ -0,0 +1,497 @@
+	}
+	return nil
+}
+
+func runCompaction(
+	ctx context.Context,
+	p sql.JobExecContext,
+	encryption *jobspb.BackupEncryptionOptions,
+	entries chan execinfrapb.RestoreSpanEntry,
+	details jobspb.BackupDetails,
+	manifest *backuppb.BackupManifest,
+	progCh chan execinfrapb.RemoteProducerMetadata_BulkProcessorProgress,
+	store cloud.ExternalStorage,
+) error {
+	defer close(progCh)
+	var encryptionOptions *kvpb.FileEncryptionOptions
+	if encryption != nil {
+		encryptionOptions = &kvpb.FileEncryptionOptions{Key: encryption.Key}
+	}
+	sinkConf := sstSinkConf{
+		id:       p.ExecCfg().DistSQLSrv.NodeID.SQLInstanceID(),
+		enc:      encryptionOptions,
+		progCh:   progCh,
+		settings: &p.ExecCfg().Settings.SV,
+	}
+	sink := makeFileSSTSink(sinkConf, store, nil)
+	sink.elideMode = manifest.ElidedPrefix
+	defer func() {
+		if err := sink.flush(ctx); err != nil {
+			log.Warningf(ctx, "failed to flush sink: %v", err)
+		}
+		logClose(ctx, sink, "SST sink")
+	}()
+	for {
+		select {
+		case entry, ok := <-entries:
+			if !ok {
+				return nil
+			}
+
+			sstIter, err := openSSTs(ctx, p, entry, encryption, details)
+			if err != nil {
+				return errors.Wrap(err, "opening SSTs")
+			}
+
+			if err := processSpanEntry(ctx, sstIter, sink, manifest); err != nil {
+				return errors.Wrap(err, "processing span entry")
+			}
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+}
+
+func processSpanEntry(
+	ctx context.Context, sstIter mergedSST, sink *fileSSTSink, manifest *backuppb.BackupManifest,
+) error {
+	defer sstIter.cleanup()
+	entry := sstIter.entry
+	startKeyMVCC, endKeyMVCC := storage.MVCCKey{Key: entry.Span.Key}, storage.MVCCKey{Key: entry.Span.EndKey}
+	prefix, err := elidedPrefix(entry.Span.Key, entry.ElidedPrefix)
+	if err != nil {
+		return err
+	}
+	if prefix != nil {
+		startKeyMVCC.Key = bytes.TrimPrefix(startKeyMVCC.Key, prefix)
+		endKeyMVCC.Key = bytes.TrimPrefix(endKeyMVCC.Key, prefix)
+	}
+	iter := sstIter.iter
+	for iter.SeekGE(startKeyMVCC); ; iter.NextKey() {
+		ok, err := iter.Valid()
+		if err != nil {
+			return err
+		}
+		if !ok {
+			break
+		}
+		key := iter.UnsafeKey()
+		if !key.Less(endKeyMVCC) {
+			break
+		}
+		uv, err := iter.UnsafeValue()
+		v, err := storage.DecodeMVCCValueAndErr(uv, err)
+		if err != nil {
+			return err
+		}
+		if err := sink.writeKey(ctx, prefix, key, v, manifest.StartTime, manifest.EndTime); err != nil {
+			return err
+		}
+	}
+	return sink.flushFile(ctx)
+}
+
+func openSSTs(
+	ctx context.Context,
+	p sql.JobExecContext,
+	entry execinfrapb.RestoreSpanEntry,
+	encryption *jobspb.BackupEncryptionOptions,
+	details jobspb.BackupDetails,
+) (mergedSST, error) {
+	var dirs []cloud.ExternalStorage
+
+	defer func() {
+		for _, dir := range dirs {
+			if err := dir.Close(); err != nil {
+				log.Warningf(ctx, "close export storage failed: %v", err)
+			}
+		}
+	}()
+
+	getIter := func(iter storage.SimpleMVCCIterator, dirsToSend []cloud.ExternalStorage, completeUpTo hlc.Timestamp) (mergedSST, error) {
```

Comment:
i don't see why this needs to be in a closure. i realize you're just copying from restore_data_processor, and i don't understand why we have this pattern there too. i'd remove the closure here and leave a todo in restore_data_processor about whether we can remove the closure.